File: Velare Ecommerce/static/js/rider/rider_earnings.js
// Rider Earnings JavaScript

document.addEventListener('DOMContentLoaded', function() {
    initializeEarnings();
    setupEventListeners();
});

// Initialize earnings page
function initializeEarnings() {
    loadEarningsData();
    loadAvailableBalance();
    loadWithdrawalHistory();
}

// Setup event listeners
function setupEventListeners() {
    // Date range inputs
    const startDate = document.getElementById('startDate');
    const endDate = document.getElementById('endDate');
    
    if (startDate) {
        startDate.addEventListener('change', handleDateRangeChange);
    }
    if (endDate) {
        endDate.addEventListener('change', handleDateRangeChange);
    }

    // Refresh button
    const refreshBtn = document.getElementById('refreshBtn');
    if (refreshBtn) {
        refreshBtn.addEventListener('click', refreshEarnings);
    }

    // Withdrawal button
    const withdrawBtn = document.getElementById('withdrawBtn');
    if (withdrawBtn) {
        withdrawBtn.addEventListener('click', handleWithdrawClick);
    }

    // Withdrawal modal buttons
    const withdrawalForm = document.getElementById('withdrawalForm');
    const closeModalBtn = document.getElementById('closeWithdrawalModal');
    const cancelWithdrawalBtn = document.getElementById('cancelWithdrawalBtn');
    const withdrawalOverlay = document.getElementById('withdrawal-modal-overlay');
    
    if (withdrawalForm) {
        withdrawalForm.addEventListener('submit', handleWithdrawalSubmit);
    }
    if (closeModalBtn) {
        closeModalBtn.addEventListener('click', closeWithdrawalModal);
    }
    if (cancelWithdrawalBtn) {
        cancelWithdrawalBtn.addEventListener('click', closeWithdrawalModal);
    }
    if (withdrawalOverlay) {
        withdrawalOverlay.addEventListener('click', closeWithdrawalModal);
    }

    // Amount input validation
    const amountInput = document.getElementById('withdrawalAmount');
    if (amountInput) {
        amountInput.addEventListener('input', validateWithdrawalAmount);
    }

    // Delivery Management toggle
    const deliveryManagementToggle = document.getElementById('deliveryManagementToggle');
    if (deliveryManagementToggle) {
        deliveryManagementToggle.addEventListener('click', toggleDeliverySubmenu);
    }

    // Navigation items (excluding logout since it's handled by logout-modal.js)
    const navItems = document.querySelectorAll('.nav-item:not(#deliveryManagementToggle):not(#logout-link)');
    navItems.forEach(item => {
        item.addEventListener('click', handleNavigation);
    });

    // Navigation subitems
    const navSubitems = document.querySelectorAll('.nav-subitem');
    navSubitems.forEach(item => {
        item.addEventListener('click', handleSubNavigation);
    });
}

// Load earnings data
function loadEarningsData(startDate = null, endDate = null) {
    console.log('=== Loading earnings with date range:', startDate, endDate);
    
    // Build query parameters
    let url = '/rider/earnings/api/earnings-data';
    if (startDate && endDate) {
        url += `?startDate=${startDate}&endDate=${endDate}`;
    }
    
    // Fetch data from server
    fetch(url)
        .then(response => {
            console.log('Response status:', response.status);
            return response.json();
        })
        .then(data => {
            console.log('Earnings data received:', data);
            
            if (data.success) {
                console.log('Number of earnings:', data.earnings.length);
                console.log('Total earnings:', data.totalEarnings);
                displayEarnings(data.earnings);
                updateTotalEarnings(data.earnings, data.totalEarnings);
            } else {
                console.error('Error loading earnings:', data.error);
                console.error('Error details:', data.details);
                alert('Error loading earnings: ' + data.error + '\n\nCheck Flask console for details.');
                displayEarnings([]);
                updateTotalEarnings([], 0);
            }
        })
        .catch(error => {
            console.error('Error fetching earnings data:', error);
            console.error('Error type:', error.name);
            console.error('Error message:', error.message);
            alert('Failed to load earnings.\n\nError: ' + error.message + '\n\nMake sure Flask server is running on http://127.0.0.1:5000');
            displayEarnings([]);
            updateTotalEarnings([], 0);
        });
}

// Filter earnings by date
function filterEarningsByDate(earnings, filter) {
    const today = new Date('2025-10-02'); // Using the current date from context
    
    switch(filter) {
        case 'today':
            return earnings.filter(e => {
                const deliveryDate = new Date(e.buyer.deliveryDate);
                return deliveryDate.toDateString() === today.toDateString();
            });
        
        case 'week':
            const weekAgo = new Date(today);
            weekAgo.setDate(today.getDate() - 7);
            return earnings.filter(e => {
                const deliveryDate = new Date(e.buyer.deliveryDate);
                return deliveryDate >= weekAgo && deliveryDate <= today;
            });
        
        case 'month':
            const monthAgo = new Date(today);
            monthAgo.setMonth(today.getMonth() - 1);
            return earnings.filter(e => {
                const deliveryDate = new Date(e.buyer.deliveryDate);
                return deliveryDate >= monthAgo && deliveryDate <= today;
            });
        
        case 'all':
        default:
            return earnings;
    }
}

// Display earnings in table
function displayEarnings(earnings) {
    const tableBody = document.getElementById('earningsTableBody');
    const emptyState = document.getElementById('emptyState');
    const table = document.querySelector('.earnings-table');
    
    if (!tableBody || !emptyState || !table) return;
    
    // Always show table headers
    table.style.display = 'table';
    
    if (earnings.length === 0) {
        // Clear table body and show empty state
        tableBody.innerHTML = '';
        emptyState.classList.add('show');
    } else {
        // Show table with data and hide empty state
        emptyState.classList.remove('show');
        tableBody.innerHTML = earnings.map(earning => createEarningRow(earning)).join('');
    }
}

// Create earning row HTML
function createEarningRow(earning) {
    return `
        <tr>
            <td>
                <span class="order-id">${earning.orderId}</span>
            </td>
            <td>
                <div class="buyer-info">
                    <div class="info-row">
                        <span class="info-label">Name:</span>
                        <span class="info-value">${earning.buyer.name}</span>
                    </div>
                    <div class="info-row">
                        <span class="info-label">Address:</span>
                        <span class="info-value">${earning.buyer.address}</span>
                    </div>
                    <div class="info-row">
                        <span class="info-label">Date:</span>
                        <span class="info-value">${formatDate(earning.buyer.deliveryDate)}</span>
                    </div>
                </div>
            </td>
            <td>
                <div class="seller-info">
                    <div class="info-row">
                        <span class="info-label">Seller:</span>
                        <span class="info-value">${earning.seller.name}</span>
                    </div>
                    <div class="info-row">
                        <span class="info-label">Store:</span>
                        <span class="info-value">${earning.seller.storeName}</span>
                    </div>
                </div>
            </td>
            <td>
                <span class="product-name">${earning.productName}</span>
            </td>
            <td>
                <div class="amount-cell">
                    <span class="amount-plus">+</span>
                    <span class="amount-value">₱${earning.amount.toFixed(2)}</span>
                </div>
            </td>
        </tr>
    `;
}

// Format date
function formatDate(dateString) {
    const date = new Date(dateString);
    const options = { year: 'numeric', month: 'short', day: 'numeric' };
    return date.toLocaleDateString('en-US', options);
}

// Update total earnings
function updateTotalEarnings(earnings, totalEarnings = null) {
    const totalEarningsElement = document.getElementById('totalEarnings');
    
    if (!totalEarningsElement) return;
    
    // Use provided totalEarnings or calculate from earnings array
    const total = totalEarnings !== null ? totalEarnings : earnings.reduce((sum, earning) => sum + earning.amount, 0);
    totalEarningsElement.textContent = `₱${total.toFixed(2)}`;
}

// Handle date range change
function handleDateRangeChange() {
    const startDate = document.getElementById('startDate')?.value;
    const endDate = document.getElementById('endDate')?.value;
    
    if (startDate && endDate) {
        console.log('Date range changed:', startDate, 'to', endDate);
        loadEarningsData(startDate, endDate);
    }
}

// Refresh earnings
function refreshEarnings() {
    console.log('Refreshing earnings...');
    const startDate = document.getElementById('startDate')?.value;
    const endDate = document.getElementById('endDate')?.value;
    loadEarningsData(startDate, endDate);
    
    // Add visual feedback
    const refreshBtn = document.getElementById('refreshBtn');
    if (refreshBtn) {
        const icon = refreshBtn.querySelector('i');
        if (icon) {
            icon.style.animation = 'spin 0.5s linear';
            setTimeout(() => {
                icon.style.animation = '';
            }, 500);
        }
    }
}

// Toggle delivery submenu with animation
function toggleDeliverySubmenu(e) {
    e.preventDefault();
    
    // Automatically navigate to List for Pickup page
    window.location.href = '/rider/pickup';
}

// Handle navigation
function handleNavigation(e) {
    const tab = e.currentTarget.getAttribute('data-tab');
    const href = e.currentTarget.getAttribute('href');
    
    console.log('Navigating to:', tab, 'href:', href);
    
    // Handle specific navigation
    if (href && href !== '#') {
        // Allow normal navigation for links with valid href
        console.log('Following href:', href);
        window.location.href = href;
        return;
    }
    
    // Prevent default for other cases
    e.preventDefault();
    
    // Handle specific navigation for items without href
    switch(tab) {
        case 'logout':
            handleLogout();
            break;
        default:
            console.log('Unknown tab:', tab);
    }
}

// Handle sub-navigation
function handleSubNavigation(e) {
    const tab = e.currentTarget.getAttribute('data-tab');
    const href = e.currentTarget.getAttribute('href');
    
    console.log('Navigating to sub-tab:', tab, 'href:', href);
    
    // Handle specific navigation
    if (href && href !== '#') {
        // Allow normal navigation for links with valid href
        console.log('Following href:', href);
        window.location.href = href;
        return;
    }
    
    // Prevent default for other cases
    e.preventDefault();
    
    // Remove active class from all subitems
    document.querySelectorAll('.nav-subitem').forEach(item => {
        item.classList.remove('active');
    });
    
    // Add active class to clicked subitem
    e.currentTarget.classList.add('active');
    
    // Handle specific sub-navigation for items without href
    switch(tab) {
        case 'list-pickup':
            window.location.href = '/rider/pickup';
            break;
        case 'active-delivery':
            window.location.href = '/rider/active-delivery';
            break;
        default:
            console.log('Unknown sub-tab:', tab);
    }
}

// Handle logout
function handleLogout() {
    // Logout is now handled by logout-modal.js
}

// Load available balance
function loadAvailableBalance() {
    fetch('/rider/earnings/api/available-balance')
        .then(response => response.json())
        .then(data => {
            if (data.success) {
                const balanceElement = document.getElementById('availableBalance');
                if (balanceElement) {
                    balanceElement.textContent = `₱${data.availableBalance.toFixed(2)}`;
                }
                
                // Store pending status for withdrawal button
                window.riderBalanceData = {
                    availableBalance: data.availableBalance,
                    hasPending: data.hasPending,
                    pendingAmount: data.pendingAmount
                };
            }
        })
        .catch(error => {
            console.error('Error loading available balance:', error);
        });
}

// Load withdrawal history
function loadWithdrawalHistory() {
    fetch('/rider/earnings/api/withdrawal-history')
        .then(response => response.json())
        .then(data => {
            if (data.success) {
                displayWithdrawalHistory(data.withdrawals);
            }
        })
        .catch(error => {
            console.error('Error loading withdrawal history:', error);
        });
}

// Display withdrawal history
function displayWithdrawalHistory(withdrawals) {
    const tableBody = document.getElementById('withdrawalHistoryBody');
    const emptyState = document.getElementById('emptyWithdrawalState');
    const table = document.querySelector('.withdrawal-table');
    
    if (!tableBody || !emptyState || !table) return;
    
    table.style.display = 'table';
    
    if (withdrawals.length === 0) {
        tableBody.innerHTML = '';
        emptyState.classList.add('show');
    } else {
        emptyState.classList.remove('show');
        tableBody.innerHTML = withdrawals.map(withdrawal => `
            <tr>
                <td>${formatDate(withdrawal.requestedAt)}</td>
                <td>₱${withdrawal.amount.toFixed(2)}</td>
                <td>
                    <span class="status-badge status-${withdrawal.status.toLowerCase()}">
                        ${withdrawal.status.charAt(0).toUpperCase() + withdrawal.status.slice(1)}
                    </span>
                </td>
                <td>${withdrawal.processedAt ? formatDate(withdrawal.processedAt) : '-'}</td>
            </tr>
        `).join('');
    }
}

// Handle withdraw button click
function handleWithdrawClick() {
    // Check if there's a pending withdrawal
    if (window.riderBalanceData && window.riderBalanceData.hasPending) {
        showWithdrawConfirmModal(
            'Pending Withdrawal',
            'You already have a pending withdrawal request.\n\nPending Amount: ₱' + window.riderBalanceData.pendingAmount.toFixed(2) + '\n\nPlease wait for it to be processed before submitting a new request.',
            null
        );
        return;
    }
    
    // Get current available balance
    const balanceElement = document.getElementById('availableBalance');
    const balanceText = balanceElement ? balanceElement.textContent : '₱0.00';
    const balance = parseFloat(balanceText.replace('₱', '').replace(',', ''));
    
    // Check if balance is valid
    if (balance <= 0) {
        showWithdrawConfirmModal(
            'Insufficient Balance',
            'Your current available balance: ' + balanceText,
            null
        );
        return;
    }
    
    // Check minimum withdrawal amount
    if (balance < 100) {
        showWithdrawConfirmModal(
            'Minimum Amount Required',
            'Minimum withdrawal amount is ₱100.00\n\nYour current available balance: ' + balanceText,
            null
        );
        return;
    }
    
    // Update modal balance display
    const modalBalance = document.getElementById('modalAvailableBalance');
    if (modalBalance) {
        modalBalance.textContent = balanceText;
    }
    
    // Set max attribute on amount input
    const amountInput = document.getElementById('withdrawalAmount');
    if (amountInput) {
        amountInput.setAttribute('max', balance);
        amountInput.value = ''; // Clear previous value
    }
    
    // Reset form
    const form = document.getElementById('withdrawalForm');
    if (form) {
        form.reset();
    }
    
    openWithdrawalModal();
}

// Open withdrawal modal
function openWithdrawalModal() {
    const modal = document.getElementById('withdrawal-modal');
    const overlay = document.getElementById('withdrawal-modal-overlay');
    
    if (modal && overlay) {
        modal.classList.add('show');
        overlay.classList.add('show');
        document.body.style.overflow = 'hidden'; // Prevent background scroll
    }
}

// Close withdrawal modal
function closeWithdrawalModal() {
    const modal = document.getElementById('withdrawal-modal');
    const overlay = document.getElementById('withdrawal-modal-overlay');
    
    if (modal && overlay) {
        modal.classList.remove('show');
        overlay.classList.remove('show');
        document.body.style.overflow = ''; // Restore scroll
    }
    
    // Reset form
    const form = document.getElementById('withdrawalForm');
    if (form) {
        form.reset();
    }
}

// Validate withdrawal amount
function validateWithdrawalAmount(e) {
    const input = e.target;
    const value = parseFloat(input.value);
    const max = parseFloat(input.getAttribute('max'));
    const min = parseFloat(input.getAttribute('min'));
    
    // Remove any existing error message
    let errorMsg = input.parentElement.querySelector('.error-message');
    if (errorMsg) {
        errorMsg.remove();
    }
    
    // Validate
    if (value < min) {
        showInputError(input, `Minimum withdrawal is ₱${min.toFixed(2)}`);
    } else if (value > max) {
        showInputError(input, `Maximum withdrawal is ₱${max.toFixed(2)}`);
        input.value = max;
    }
}

// Show input error
function showInputError(input, message) {
    let errorMsg = input.parentElement.querySelector('.error-message');
    if (!errorMsg) {
        errorMsg = document.createElement('small');
        errorMsg.className = 'error-message';
        input.parentElement.appendChild(errorMsg);
    }
    errorMsg.textContent = message;
}

// Handle withdrawal form submit
function handleWithdrawalSubmit(e) {
    e.preventDefault();
    
    const form = e.target;
    const formData = new FormData(form);
    
    const amount = parseFloat(formData.get('amount'));
    const method = formData.get('method');
    const notes = formData.get('notes');
    
    // Get available balance
    const balanceElement = document.getElementById('availableBalance');
    const balanceText = balanceElement ? balanceElement.textContent : '₱0.00';
    const availableBalance = parseFloat(balanceText.replace('₱', '').replace(',', ''));
    
    // Final validation
    if (isNaN(amount) || amount <= 0) {
        showWithdrawConfirmModal('Invalid Amount', 'Please enter a valid amount', null);
        return;
    }
    
    if (amount < 100) {
        showWithdrawConfirmModal('Minimum Amount Required', 'Minimum withdrawal amount is ₱100.00', null);
        return;
    }
    
    if (amount > availableBalance) {
        showWithdrawConfirmModal(
            'Insufficient Balance',
            `Requested: ₱${amount.toFixed(2)}\nAvailable: ₱${availableBalance.toFixed(2)}`,
            null
        );
        return;
    }
    
    // Check if withdrawal would result in negative balance
    if (availableBalance - amount < 0) {
        showWithdrawConfirmModal(
            'Invalid Withdrawal',
            'This withdrawal would result in a negative balance.\n\nPlease check your available balance.',
            null
        );
        return;
    }
    
    // Disable submit button
    const submitBtn = form.querySelector('button[type="submit"]');
    if (submitBtn) {
        submitBtn.disabled = true;
        submitBtn.innerHTML = '<i class="bi bi-hourglass-split"></i> Processing...';
    }
    
    // Send withdrawal request
    fetch('/rider/earnings/api/withdraw', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ 
            amount: amount,
            method: method,
            notes: notes
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            showNotification('Withdrawal request submitted successfully! Amount: ₱' + amount.toFixed(2), 'success');
            closeWithdrawalModal();
            // Reload data
            loadAvailableBalance();
            loadWithdrawalHistory();
        } else {
            showNotification(data.error, 'error');
        }
    })
    .catch(error => {
        console.error('Error submitting withdrawal:', error);
        showNotification('Failed to submit withdrawal request', 'error');
    })
    .finally(() => {
        // Re-enable submit button
        if (submitBtn) {
            submitBtn.disabled = false;
            submitBtn.innerHTML = '<i class="bi bi-check-circle"></i> Submit Request';
        }
    });
}

// Notification function
function showNotification(message, type = 'success') {
    // Remove any existing notifications first
    const existingNotifications = document.querySelectorAll('.notification');
    existingNotifications.forEach(notif => {
        notif.style.animation = 'slideOut 0.3s ease';
        setTimeout(() => {
            notif.remove();
        }, 300);
    });
    
    // Create new notification
    const notification = document.createElement('div');
    notification.className = `notification notification-${type}`;
    notification.innerHTML = `
        <i class="bi bi-${type === 'success' ? 'check-circle' : type === 'error' ? 'x-circle' : 'info-circle'}"></i>
        <span>${message}</span>
    `;
    
    notification.style.cssText = `
        position: fixed;
        top: 100px;
        right: 30px;
        padding: 16px 24px;
        background: ${type === 'success' ? '#ecfdf5' : type === 'error' ? '#fef2f2' : '#eff6ff'};
        color: ${type === 'success' ? '#059669' : type === 'error' ? '#dc2626' : '#2563eb'};
        border: 1.5px solid ${type === 'success' ? '#10b981' : type === 'error' ? '#f87171' : '#60a5fa'};
        border-radius: 8px;
        box-shadow: 0 4px 16px rgba(0,0,0,0.1);
        z-index: 10000;
        display: flex;
        align-items: center;
        gap: 12px;
        font-family: 'Montserrat', sans-serif;
        font-size: 0.95rem;
        font-weight: 600;
        animation: slideIn 0.3s ease;
    `;
    
    document.body.appendChild(notification);
    
    setTimeout(() => {
        notification.style.animation = 'slideOut 0.3s ease';
        setTimeout(() => {
            notification.remove();
        }, 300);
    }, 3000);
}

// Show withdrawal confirmation modal (logout style)
function showWithdrawConfirmModal(title, message, onConfirm) {
    const modal = document.getElementById('withdraw-confirm-modal');
    const overlay = document.getElementById('withdraw-confirm-modal-overlay');
    const titleElement = document.getElementById('withdrawConfirmTitle');
    const messageElement = document.getElementById('withdrawConfirmMessage');
    const confirmBtn = document.getElementById('confirm-withdraw');
    const cancelBtn = document.getElementById('cancel-withdraw');
    
    if (!modal || !overlay) return;
    
    // Set title and message
    if (titleElement) titleElement.textContent = title;
    if (messageElement) messageElement.textContent = message;
    
    // Show/hide confirm button based on whether there's a callback
    if (confirmBtn) {
        if (onConfirm) {
            confirmBtn.style.display = 'inline-block';
            // Remove old listeners
            const newConfirmBtn = confirmBtn.cloneNode(true);
            confirmBtn.parentNode.replaceChild(newConfirmBtn, confirmBtn);
            // Add new listener
            newConfirmBtn.addEventListener('click', () => {
                closeWithdrawConfirmModal();
                onConfirm();
            });
        } else {
            confirmBtn.style.display = 'none';
        }
    }
    
    // Cancel button closes modal
    if (cancelBtn) {
        const newCancelBtn = cancelBtn.cloneNode(true);
        cancelBtn.parentNode.replaceChild(newCancelBtn, cancelBtn);
        newCancelBtn.addEventListener('click', closeWithdrawConfirmModal);
        // If no confirm button, change text to "OK"
        if (!onConfirm) {
            newCancelBtn.textContent = 'OK';
        } else {
            newCancelBtn.textContent = 'Cancel';
        }
    }
    
    // Show modal
    modal.classList.add('active');
    overlay.classList.add('show');
    document.body.style.overflow = 'hidden';
}

// Close withdrawal confirmation modal
function closeWithdrawConfirmModal() {
    const modal = document.getElementById('withdraw-confirm-modal');
    const overlay = document.getElementById('withdraw-confirm-modal-overlay');
    
    if (modal && overlay) {
        modal.classList.remove('active');
        overlay.classList.remove('show');
        document.body.style.overflow = '';
    }
}

// Add CSS animations
const style = document.createElement('style');
style.textContent = `
    @keyframes spin {
        from { transform: rotate(0deg); }
        to { transform: rotate(360deg); }
    }
    
    @keyframes slideIn {
        from {
            transform: translateX(400px);
            opacity: 0;
        }
        to {
            transform: translateX(0);
            opacity: 1;
        }
    }
    
    @keyframes slideOut {
        from {
            transform: translateX(0);
            opacity: 1;
        }
        to {
            transform: translateX(400px);
            opacity: 0;
        }
    }
`;
document.head.appendChild(style);

File: Velare Ecommerce/blueprints/rider_earnings.py
from flask import Blueprint, render_template, session, redirect, url_for, jsonify, request
import sys
import os
from datetime import datetime, timedelta

# Add the parent directory to the path to import db_config
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from database.supabase_helper import *

rider_earnings_bp = Blueprint('rider_earnings', __name__)

@rider_earnings_bp.route('/rider/earnings')
def rider_earnings():
    # Check if user is logged in and is a rider
    if 'user_id' not in session or session.get('user_type') != 'rider':
        return redirect(url_for('auth.login'))
    
    try:
        # Get rider information from Supabase
        rider_data = get_rider_by_user_id(session['user_id'])
        
        if not rider_data:
            return render_template('rider/rider_earnings.html', error='Rider profile not found')
        
        return render_template('rider/rider_earnings.html', rider=rider_data)
        
    except Exception as e:
        print(f"Error fetching rider data: {e}")
        return render_template('rider/rider_earnings.html', error='Failed to load rider data')

@rider_earnings_bp.route('/rider/earnings/api/earnings-data', methods=['GET'])
def get_earnings_data():
    """API endpoint to fetch earnings data for the current rider"""
    if 'user_id' not in session or session.get('user_type') != 'rider':
        return jsonify({'error': 'Unauthorized'}), 401
    
    start_date = request.args.get('startDate')
    end_date = request.args.get('endDate')
    
    try:
        # Get rider_id for the current user
        rider = get_rider_by_user_id(session['user_id'])
        
        if not rider:
            return jsonify({'error': 'Rider not found'}), 404
        
        rider_id = rider['rider_id']
        
        # Fetch earnings data from Supabase
        earnings_data = get_rider_earnings_history(rider_id, start_date, end_date)
        
        # Debug logging
        print(f"=== EARNINGS DEBUG ===")
        print(f"Rider ID: {rider_id}")
        print(f"Date Range: {start_date} to {end_date}")
        print(f"Number of earnings found: {len(earnings_data)}")
        
        # Calculate total earnings
        total_earnings = sum(float(e.get('rider_earnings', 0) or 0) for e in earnings_data)
        
        # Format the data for frontend
        supabase = get_supabase()
        formatted_earnings = []
        
        for earning in earnings_data:
            order = earning.get('orders', {})
            buyer = order.get('buyers', {})
            seller = order.get('sellers', {})
            
            # Get address
            address_id = order.get('address_id')
            buyer_address = 'N/A'
            if address_id:
                addr_response = supabase.table('addresses').select('full_address').eq('address_id', address_id).execute()
                if addr_response.data:
                    buyer_address = addr_response.data[0].get('full_address', 'N/A')
            
            # Get product names from order items
            order_items_response = supabase.table('order_items').select('''
                products (
                    product_name
                )
            ''').eq('order_id', earning['order_id']).execute()
            
            product_names = []
            if order_items_response.data:
                for item in order_items_response.data:
                    product = item.get('products', {})
                    if product and product.get('product_name'):
                        product_names.append(product['product_name'])
            
            product_name = ', '.join(product_names) if product_names else 'N/A'
            
            # Parse delivered_at
            delivered_date = None
            if earning.get('delivered_at'):
                try:
                    delivered_date = datetime.fromisoformat(earning['delivered_at'].replace('Z', '+00:00')).strftime('%Y-%m-%d')
                except:
                    delivered_date = None
            
            formatted_earnings.append({
                'orderId': order.get('order_number'),
                'buyer': {
                    'name': f"{buyer.get('first_name', '')} {buyer.get('last_name', '')}",
                    'address': buyer_address,
                    'deliveryDate': delivered_date
                },
                'seller': {
                    'name': f"{seller.get('first_name', '')} {seller.get('last_name', '')}",
                    'storeName': seller.get('shop_name')
                },
                'productName': product_name,
                'amount': float(earning.get('rider_earnings', 0) or 0)
            })
        
        return jsonify({
            'success': True, 
            'earnings': formatted_earnings,
            'totalEarnings': total_earnings
        }), 200
        
    except Exception as e:
        import traceback
        error_details = traceback.format_exc()
        print(f"=== ERROR FETCHING EARNINGS ===")
        print(f"Error: {e}")
        print(f"Traceback: {error_details}")
        return jsonify({
            'success': False,
            'error': f'Failed to fetch earnings data: {str(e)}',
            'details': error_details
        }), 500


@rider_earnings_bp.route('/rider/earnings/api/available-balance', methods=['GET'])
def get_available_balance():
    """API endpoint to get rider's available balance for withdrawal"""
    if 'user_id' not in session or session.get('user_type') != 'rider':
        return jsonify({'error': 'Unauthorized'}), 401
    
    try:
        # Get rider_id
        rider = get_rider_by_user_id(session['user_id'])
        
        if not rider:
            return jsonify({'error': 'Rider not found'}), 404
        
        rider_id = rider['rider_id']
        
        supabase = get_supabase()
        
        # Check for pending withdrawal
        pending_response = supabase.table('rider_withdrawals').select('amount', count='exact').eq('rider_id', rider_id).eq('status', 'pending').execute()
        has_pending = (pending_response.count or 0) > 0
        pending_amount = sum(float(w.get('amount', 0)) for w in (pending_response.data or []))
        
        # Calculate total earnings from delivered orders
        total_earnings = get_rider_total_earnings(rider_id)
        
        # Calculate total withdrawn (completed withdrawals only)
        withdrawn_response = supabase.table('rider_withdrawals').select('amount').eq('rider_id', rider_id).eq('status', 'completed').execute()
        total_withdrawn = sum(float(w.get('amount', 0)) for w in (withdrawn_response.data or []))
        
        # Calculate available balance (subtract both completed and pending withdrawals)
        available_balance = total_earnings - total_withdrawn - pending_amount
        
        return jsonify({
            'success': True,
            'availableBalance': available_balance,
            'totalEarnings': total_earnings,
            'totalWithdrawn': total_withdrawn,
            'pendingAmount': pending_amount,
            'hasPending': has_pending
        }), 200
        
    except Exception as e:
        print(f"Error getting available balance: {e}")
        return jsonify({'error': str(e)}), 500


@rider_earnings_bp.route('/rider/earnings/api/withdraw', methods=['POST'])
def request_withdrawal():
    """API endpoint to request a withdrawal"""
    if 'user_id' not in session or session.get('user_type') != 'rider':
        return jsonify({'error': 'Unauthorized'}), 401
    
    data = request.get_json()
    amount = data.get('amount')
    method = data.get('method', 'Cash')
    notes = data.get('notes', '')
    
    # Validation
    if not amount or float(amount) <= 0:
        return jsonify({'error': 'Invalid amount'}), 400
    
    amount = float(amount)
    
    # Minimum withdrawal check
    if amount < 100:
        return jsonify({'error': 'Minimum withdrawal amount is ₱100.00'}), 400
    
    try:
        # Get rider_id
        rider = get_rider_by_user_id(session['user_id'])
        
        if not rider:
            return jsonify({'error': 'Rider not found'}), 404
        
        rider_id = rider['rider_id']
        
        supabase = get_supabase()
        
        # Check for existing pending withdrawal
        pending_response = supabase.table('rider_withdrawals').select('withdrawal_id', count='exact').eq('rider_id', rider_id).eq('status', 'pending').execute()
        
        if (pending_response.count or 0) > 0:
            return jsonify({
                'error': 'You already have a pending withdrawal request. Please wait for it to be processed before submitting a new one.'
            }), 400
        
        # Calculate available balance
        total_earnings = get_rider_total_earnings(rider_id)
        
        # Calculate total withdrawn (completed withdrawals only)
        withdrawn_response = supabase.table('rider_withdrawals').select('amount').eq('rider_id', rider_id).eq('status', 'completed').execute()
        total_withdrawn = sum(float(w.get('amount', 0)) for w in (withdrawn_response.data or []))
        
        # Calculate pending withdrawals
        pending_amount_response = supabase.table('rider_withdrawals').select('amount').eq('rider_id', rider_id).eq('status', 'pending').execute()
        pending_amount = sum(float(w.get('amount', 0)) for w in (pending_amount_response.data or []))
        
        # Available balance = total earnings - completed withdrawals - pending withdrawals
        available_balance = total_earnings - total_withdrawn - pending_amount
        
        # Check if sufficient balance
        if amount > available_balance:
            return jsonify({
                'error': f'Insufficient balance. Available: ₱{available_balance:.2f}'
            }), 400
        
        # Additional safety check to prevent negative balance
        if available_balance - amount < 0:
            return jsonify({
                'error': 'This withdrawal would result in a negative balance. Please check your available balance.'
            }), 400
        
        # Insert withdrawal request
        supabase.table('rider_withdrawals').insert({
            'rider_id': rider_id,
            'amount': amount,
            'withdrawal_method': method,
            'status': 'pending',
            'notes': notes
        }).execute()
        
        return jsonify({
            'success': True,
            'message': 'Withdrawal request submitted successfully',
            'newBalance': available_balance - amount
        }), 200
        
    except Exception as e:
        print(f"Error requesting withdrawal: {e}")
        return jsonify({'error': str(e)}), 500


@rider_earnings_bp.route('/rider/earnings/api/withdrawal-history', methods=['GET'])
def get_withdrawal_history():
    """API endpoint to get withdrawal history"""
    if 'user_id' not in session or session.get('user_type') != 'rider':
        return jsonify({'error': 'Unauthorized'}), 401
    
    try:
        # Get rider_id
        rider = get_rider_by_user_id(session['user_id'])
        
        if not rider:
            return jsonify({'error': 'Rider not found'}), 404
        
        rider_id = rider['rider_id']
        
        # Fetch withdrawal history from Supabase
        supabase = get_supabase()
        response = supabase.table('rider_withdrawals').select('''
            withdrawal_id,
            amount,
            withdrawal_method,
            status,
            requested_at,
            processed_at,
            notes
        ''').eq('rider_id', rider_id).order('requested_at', desc=True).execute()
        
        withdrawals = clean_supabase_data(response.data) if response.data else []
        
        # Format data
        formatted_withdrawals = []
        for w in withdrawals:
            # Parse dates
            requested_at = None
            processed_at = None
            
            if w.get('requested_at'):
                try:
                    requested_at = datetime.fromisoformat(w['requested_at'].replace('Z', '+00:00')).strftime('%Y-%m-%d %H:%M:%S')
                except:
                    requested_at = None
            
            if w.get('processed_at'):
                try:
                    processed_at = datetime.fromisoformat(w['processed_at'].replace('Z', '+00:00')).strftime('%Y-%m-%d %H:%M:%S')
                except:
                    processed_at = None
            
            formatted_withdrawals.append({
                'withdrawalId': w['withdrawal_id'],
                'amount': float(w['amount']),
                'method': w['withdrawal_method'],
                'status': w['status'],
                'requestedAt': requested_at,
                'processedAt': processed_at,
                'notes': w.get('notes')
            })
        
        return jsonify({
            'success': True,
            'withdrawals': formatted_withdrawals
        }), 200
        
    except Exception as e:
        print(f"Error getting withdrawal history: {e}")
        return jsonify({'error': str(e)}), 500
